### ISSUE TITLE
Refactor to make it easier to add other auth providers

### ISSUE THREAD

**Issue by shankari** [e-mission/cordova-jwt-auth#17]
we want to be able to support other auth providers
In particular, we want to support:
- OpenID Connect, for @Andrew-Tan
- "dev mode" with token = user email, for testing against localhost

**Comment by shankari** [comment]
Pull request for fix is #18

**Comment by shankari** [comment]
First step is to remove all references to google from the plugins and the internal interface and come up with a standard interface that we can generalize.
These changes are in c84b93efc3e68d253e018e80edffc159246f0e94 (iOS) and eddfe9bd2bedf5f932f30ee2fcbc9de7bbecf497 (android).

Both the changes work for both initial install, upgrade, and re-launch, with one caveat.
When I test the initial install on android while connected to the debugger, sometimes, `signIn` is called twice. This causes the app to crash because the second call to 

```
cordova.setActivityResultCallback(this);
```

causes cordova to cancel the first callback with a call to `onActivityResult` with `requestCode` = `RESULT_CANCELLED`, which returns from the call with an error.

```
    public void setActivityResultCallback(CordovaPlugin plugin) {
        // Cancel any previously pending activity.
        if (activityResultCallback != null) {
            activityResultCallback.onActivityResult(activityResultRequestCode, Activity.RESULT_CANCELED, null);
        }
        activityResultCallback = plugin;
}
```
https://github.com/apache/cordova-android/blob/ad01d28351c13390aff4549258a0f06882df59f5/framework/src/org/apache/cordova/CordovaInterfaceImpl.java#L74

and then the second callback to `onActivityResult`, with the result, crashes.

This does not happen when not connected to the debugger. And when I added a javascript breakpoint to `signIn` (in chrome), it didn't happen either.

I will re-test this after making all changes, but it doesn't seem to be super important.

**Comment by shankari** [comment]
Second step is to create an actual interface (`interface` in android and `@protocol` in iOS) that defines the methods for the auth methods. Then, we create a factory to obtain the correct instance of the interface by using the connection settings. Finally, we change the external references to use the new factory and not the auth implementation directly.

Changes:
- iOS: https://github.com/e-mission/cordova-jwt-auth/pull/18/commits/268d52e46bc206ddf2b11be0f671da0c268c56e1
- android: https://github.com/e-mission/cordova-jwt-auth/pull/18/commits/847905deba2480df36d741e1a47f84af96bde3f7
- settings: https://github.com/e-mission/cordova-connection-settings/pull/7
- communication: https://github.com/e-mission/cordova-server-communication/pull/17/commits/8bd2c0cbeb2680bd668ef31e719e27cac79557a1
- sync: https://github.com/e-mission/cordova-server-sync/pull/29
- data collection: https://github.com/e-mission/e-mission-data-collection/pull/159

**Comment by shankari** [comment]
Testing done:
initial install and re-launch on both android and iOS

**Comment by shankari** [comment]
Note: need to test upgrade from store version
use TestFlight, beta channel...

**Comment by shankari** [comment]
Merging everything other than the auth changes because from this point on - all changes are going to be around adding other auth implementations.

We may need to change `uiSignIn()` and callback interfaces slightly as we experiment with the dev mode implementation, but that should only affect the plugin, no other code.

**Comment by shankari** [comment]
Next, change the `uiSignIn()` interface to pass in the plugin. This ensures that:
- we have the full plugin to make javascript calls
- we don't have to do awkward stunts like setting the activity or the controller
- we can register for the activity callback only for the google auth plugin since it is not needed for other plugins

This change is https://github.com/e-mission/cordova-jwt-auth/pull/18/commits/08952ed5c5202ea9b8db4d1144830169cd698f3f

**Comment by shankari** [comment]
Next, rename the google auth wrapper on iOS to match what it actually does instead of the generic-sounding `AuthCompletionHandler`.

Related changes are
https://github.com/e-mission/cordova-jwt-auth/pull/18/commits/f9b82d8e8a9e8583cce613cf37583e4e68fa0526, https://github.com/e-mission/cordova-jwt-auth/pull/18/commits/a2f008a11f6be0e8e336f113682bf45da69fe179, and https://github.com/e-mission/cordova-jwt-auth/pull/18/commits/7f33c0106f9a584453bc59ed3ceaf9e62d26d4a1

**Comment by shankari** [comment]
Finally, we are ready to move to the dev implementation. The high level plan is:
- execute javascript that prompts the user for their email
- return it to native code through a custom URL (since we have it defined anyway)
- save into file or keychain
- return from there forevermore

**Comment by shankari** [comment]
Ran into lots of issues while returning the email via custom URL.
Concretely, how do we open that URL programmatically?

While I initially experimented by changing the javascript launched as part of the app, this part can be tested by connecting with a debugger to the cordova app and typing commands into the console.

My first try was to use XHR, but it failed because it doesn't appear to support custom URLs. Although the documentation (https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest) says that
> Despite its name, XMLHttpRequest can be used to retrieve any type of data, not just XML, and it supports protocols other than HTTP (including file and ftp).
it doesn't look like custom protocols are supported, and even support for `ftp://` URLs is skimpy.
https://stackoverflow.com/questions/14839838/what-is-the-syntax-to-do-a-cross-domain-xmlhttprequest-to-an-ftp-server

```
> var xhr = new XMLHttpRequest();
> xhr.open("GET", "emission://auth?method='dev'&token='testToken'", true);
> xhr.send();
[Error] Failed to load resource: unsupported URL (auth?method='dev'&token='testtoken', line 0)
```

Next, I created a test page that embedded such a URL (http://cs.berkeley.edu/~shankari/test_emission_url.html) and confirmed that clicking on the URL launched the app.

So obviously the app was working - I just needed to figure out how to programmatically click on the the URL.

So I next tried `window.open` (https://developer.mozilla.org/en-US/docs/Web/API/Window/open), which should
> loads the specified resource into the browsing context (window or tab) with the specified name. If the name doesn't exist, then a new window is opened and the specified resource is loaded into its browsing context.

This has no errors, but does not do anything in a cordova webapp (webview) setting.

```
> window.open("emission://auth?method='dev'&token='testToken'")
< {channels: Object, _eventHandler: function, close: function, show: function, hide: function, …}
```

| Before          | After        |
| ------------ | --------- |
| ![simulator screen shot aug 20 2017 9 32 16 pm](https://user-images.githubusercontent.com/2423263/29503733-4261aa3a-85ef-11e7-9ce5-5715fd7ee78f.png) | ![simulator screen shot aug 20 2017 9 32 49 pm](https://user-images.githubusercontent.com/2423263/29503732-42604136-85ef-11e7-980c-5901916728d0.png) |

Then, I tried the inapp browser plugin, with the same result.

```
> cordova.InAppBrowser.open("emission://auth?method='dev'&token='testToken'")
< {channels: Object, _eventHandler: function, close: function, show: function, hide: function, …}
```

Then, I tried the inapp browser plugin, but opening in a new window, and that finally worked.

```
> cordova.InAppBrowser.open("emission://auth?method='dev'&token='testToken'", "_blank")
< {channels: Object, _eventHandler: function, close: function, show: function, hide: function, …}
```

Next, I wanted to make sure that I automatically close the IAB window after the URL is opened. I tried listening to the `loadstop` event, but it turns out that, probably because this is not really a URL, it never calls `loadstop`. `loadstart` is called, though, so we can listen to it, and close the window after a short delay.

```
> var iab = cordova.InAppBrowser.open("emission://auth?method='dev'&token='testToken'", "_blank")
iab.addEventListener('loadstart', function(event) {console.log("load of "+event.url+" started");});
iab.addEventListener('loadstop', function(event) {console.log("load of "+event.url+" stopped");});
console.log("done");
[Log] done (console-via-logger.js, line 173)
< undefined
[Log] GOT URL:emission://auth?method='dev'&token='testToken' (console-via-logger.js, line 173)
[Log] Checking to see whether we are ready to load the screen (console-via-logger.js, line 173)
[Log] DEBUG:About to set theme from preference (console-via-logger.js, line 173)
[Log] DEBUG:preferred theme = null (console-via-logger.js, line 173)
[Log] DEBUG:set dark_theme = false (console-via-logger.js, line 173)
[Log] DEBUG:About to navigate to preferred tab (console-via-logger.js, line 173)
[Log] DEBUG:in isIntroDone, read_val = true (console-via-logger.js, line 173)
[Log] DEBUG:in isIntroDone, returning true (console-via-logger.js, line 173)
[Log] load of emission://auth?method='dev'&token='testToken' started (console-via-logger.js, line 173)
[Log] DEBUG:req_consent = [object Object] curr_consented = [object Object] (console-via-logger.js, line 173)
[Log] DEBUG:resolving with empty information (console-via-logger.js, line 173)
[Log] DEBUG:in isIntroDone, read_val = true (console-via-logger.js, line 173)
[Log] DEBUG:in isIntroDone, returning true (console-via-logger.js, line 173)
[Log] DEBUG:curr_consented = [object Object]isIntroDone = true (console-via-logger.js, line 173)
[Log] DEBUG:in isIntroDone, read_val = true (console-via-logger.js, line 173)
[Log] DEBUG:in isIntroDone, returning true (console-via-logger.js, line 173)
[Log] Consented in local storage, no need to show consent (console-via-logger.js, line 173)
[Log] DEBUG:changing state to root.main.metrics (console-via-logger.js, line 173)
[Log] loading root.main.metrics (console-via-logger.js, line 173)
```

**Comment by shankari** [comment]
Also, FYI, reloading into the same window with the `_self` option doesn't work even when we have an IAB. For example:

Step 1
-----
```
var iab = cordova.InAppBrowser.open("emission://auth?method='dev'&token='testToken'", "_blank")
iab.addEventListener('loadstart', function(event) {console.log("load of "+event.url+" started");});
iab.addEventListener('loadstop', function(event) {console.log("load of "+event.url+" stopped");});
console.log("done");
```

Blank IAB, messages about load starting

```
[Log] GOT URL:emission://auth?method='dev'&token='testToken' (console-via-logger.js, line 173)
> cordova.InAppBrowser.open("https://www.nytimes.com", "_self");
[Log] load of emission://auth?method='dev'&token='testToken' started (console-via-logger.js, line 173)
```

STEP 2: Try to open canonical website with `_self`

```
> cordova.InAppBrowser.open("https://www.nytimes.com", "_self");
< {channels: Object, _eventHandler: function, close: function, show: function, hide: function, …}
```

No change, still blank iab.

STEP 3: Open a canonical website with `_blank`
nytimes shows up. Note that only one window is visible in the task manager, at least on iOS.
We'll add side-by-side screenshots once android is done

**Comment by shankari** [comment]
> Then, I tried the inapp browser plugin, but opening in a new window, and that finally worked.

Today morning, I tried the same code on android, and it didn't work. I kept getting errors saying that the URL was not supported.

```
> cordova.InAppBrowser.open("emission://auth?method=dummy-dev&token=test-android-auth", "_blank")
< InAppBrowser {channels: Object}
```

| Doesn't work |  `window.open` still doesn't work | `_system` works! |
|-------|------| ----- |
| ![screenshot_1503343843](https://user-images.githubusercontent.com/2423263/29535047-40a758c4-866d-11e7-8e94-f821b3470806.png) | ![screenshot_1503344367](https://user-images.githubusercontent.com/2423263/29535268-104dd242-866e-11e7-9576-239deebfac44.png) | ![screenshot_1503345268](https://user-images.githubusercontent.com/2423263/29535743-e92c5e7a-866f-11e7-940c-6b5715588e0f.png) |

Searching for this on the internet just found issues about issues with custom URLs from 2014, but nothing recent. Custom URLs did work from the browser, of course.

Some people did suggest customizing the handling of URLs (e.g. https://stackoverflow.com/questions/3583264/support-for-other-protocols-in-android-webview). This maps to the `onOverrideUrlLoading` in the plugin, so I overrode that, and it still didn't work.

It turns out that this was because the `onOverrideUrlLoading` was never called, as I confirmed by adding a breakpoint to the method. I suspected that this was because that was only called on the cordova webview, not on the IAB webview, so I tried `window.open`, and that didn't help either.

But then I noticed that the screen looked identical for both of them! And the logs showed that `window.open` launched an IAB after all.

```
> window.open("emission://auth?method=dummy-dev&token=test-android-auth")
< InAppBrowser {channels: Object}
```

Aha! This is because of https://cordova.apache.org/docs/en/latest/reference/cordova-plugin-inappbrowser/index.html
> For backwards compatibility, this plugin also hooks window.open. However, the plugin-installed hook of window.open can have unintended side effects (especially if this plugin is included only as a dependency of another plugin). The hook of window.open will be removed in a future major release. Until the hook is removed from the plugin, apps can manually restore the default behaviour:

Ok, so what is happening in the IAB, and can I add a hook to that webview?

Here's the current hook, and it only handles `tel`, `sms` and `geo`.

https://github.com/apache/cordova-plugin-inappbrowser/blob/master/src/android/InAppBrowser.java#L945

```
        /**
         * Override the URL that should be loaded
         *
         * This handles a small subset of all the URIs that would be encountered.
         *
         * @param webView
         * @param url
         */

        @Override
public boolean shouldOverrideUrlLoading(WebView webView, String url) {
```

and it explicitly does not allow any other URLs.
https://github.com/apache/cordova-plugin-inappbrowser/blob/master/src/android/InAppBrowser.java#L1016

I saw this message in the android console as well.
```
            {
                // Assume that everything is HTTP at this point, because if we don't specify,
                // it really should be.  Complain loudly about this!!!
                LOG.e(LOG_TAG, "Possible Uncaught/Unknown URI");
                newloc = "http://" + url;
          }
```

So it looks like it would be hard to add other hooks to the IAB. Do we need to remove the `window.open` hook and see if it works? The workaround `delete window.open` seems scary!

But wait, there's another method `openExternal` that looks a lot simpler and already does stuff with intents.
https://github.com/apache/cordova-plugin-inappbrowser/blob/master/src/android/InAppBrowser.java#L394

And it is invoked if we use the `_system` option.

```
                    else if (SYSTEM.equals(target)) {
                        LOG.d(LOG_TAG, "in system");
                        result = openExternal(url);
                    }
```

So can we just try this? Works!

```
> cordova.InAppBrowser.open("emission://auth?method=dummy-dev&token=test-android-auth", "_system")
< InAppBrowser {channels: Object}
controllers.js:24 GOT URL:emission://auth?method=dummy-dev&token=test-android-auth
startprefs.js:233 Checking to see whether we are ready to load the screen
unifiedlogger.js:49 DEBUG:About to set theme from preference
unifiedlogger.js:49 DEBUG:preferred theme = null
unifiedlogger.js:49 DEBUG:set dark_theme = false
unifiedlogger.js:49 DEBUG:About to navigate to preferred tab
unifiedlogger.js:49 DEBUG:in isIntroDone, read_val = true
unifiedlogger.js:49 DEBUG:in isIntroDone, returning true
unifiedlogger.js:49 DEBUG:req_consent = [object Object] curr_consented = [object Object]
unifiedlogger.js:49 DEBUG:resolving with empty information
unifiedlogger.js:49 DEBUG:in isIntroDone, read_val = true
unifiedlogger.js:49 DEBUG:in isIntroDone, returning true
unifiedlogger.js:49 DEBUG:curr_consented = [object Object]isIntroDone = true
unifiedlogger.js:49 DEBUG:in isIntroDone, read_val = true
unifiedlogger.js:49 DEBUG:in isIntroDone, returning true
startprefs.js:94 Consented in local storage, no need to show consent
unifiedlogger.js:49 DEBUG:changing state to root.main.metrics
startprefs.js:205 loading root.main.metrics
```

**Comment by shankari** [comment]
Ok so almost final set of changes.
First, finally add support for the dummy dev mode auth described in 
https://github.com/e-mission/cordova-jwt-auth/issues/17#issuecomment-323642937
in 4d5eeccba3d18ae03518e805193bbb9188941851

Because we have a dummy auth mode now, we can return it by default instead of failing or returning null if nothing is configured. But what if we took it one step further? 

It would be nice to be able to connect to the server out of the box without *any* configuration. In order to do this, we change the connection settings to return default values instead of null if the config file is missing. These changes are in https://github.com/e-mission/cordova-connection-settings/pull/8

So now I can revert the documentation changes from https://github.com/e-mission/e-mission-phone/commit/2b31708b285dfdef8d096e9dcbcfa52b745151b3

**Comment by shankari** [comment]
This is the final set of changes, and it is an implementation-only change. While having the javascript in the native code files was fine for the first pass, it is not a long-term solution for the following reasons.
- the strings that represent lines of code have to be quoted with `""`, which means that we have to use `''` for strings *in* the code. Combined with the lack of syntax highlighting, this was very confusing and I ran into several issues during testing with forgetting quotes
- multi-line lines are hard to see due to lack of indentation
- it is not possible to connect with a debugger to see what's going on. I had to resort to adding `alert()`s for debugging
- duplicated code meant that changes in one platform had to be copied to the other.

I fixed this by using the same technique as `dispatchIOSLocalNotification`
https://github.com/e-mission/e-mission-transition-notify/blame/master/www/transitionnotify.js

I put the javascript into the plugin code, and then simply invoked that function from the native code. The related changes are in 34318a6386074588a549330e5f9c49ee9fb9ab1b

With this change, I declare this feature enhancement complete, and plan to merge the changes ASAP
